### DISCUSSION TITLE
Migrating bsky.social to Multiple PDS Instances (Nov 2023)

### DISCUSSION THREAD

**Discussion by bnewbold** [bluesky-social/atproto#1832] [Dev Announcements]
As part of our [operational work in preparation for federation](https://github.com/bluesky-social/atproto/discussions/1711), we are starting to migrate PDS accounts away from the existing monolithic `bsky.social` instance to several `*.*.host.bsky.network` PDS instances. We have implemented a new "account entryway" service as a "virtual PDS" which will continue to operate at `bsky.social`.

These changes are expected to roll out on the main network the week of **2023-11-06** (aka, this week, starting today). Every Bluesky-hosted account will be migrated to a new PDS, which will involved updating DID PLC documents with new PDS service locations and new signing keys. The roll-out may take longer than a week.

### What *will not* change

* Client apps, bots, and other integrations can continue to connect to `bsky.social` as an API endpoint, and requests will be proxied to the appropriate endpoint. They **do not** need to connect directly to the account's PDS for now, especially for authentication (login and session refresh)
* Bluesky AppView API (`api.bsky.app`) is unaffected
* Bluesky BGS firehose is unaffected

## What *will* change:

* DID PLC documents will be updated, with account signing keys and PDS endpoints updated
* Account repos will, of course, be canonically hosted at the new hostname indicated in the DID document. repo events will be emitted from the canonical PDS host for the account, not from `bsky.social`; subscribe to the BGS firehose (`bsky.network`) to get all events from all accounts.
* Clients (apps, bots, etc) *may* connect directly to the PDS hostname instead of `bsky.social`
* Each account will now have a unique signing key
* some `bsky.social` endpoints may return 3xx redirects to the specific PDS endpoint, instead of proxying. Specifically, `com.atproto.sync.getBlob` (for downloading images and other media). Most clients and applications use the AppView-provided CDN URLs, not blobs directly.
* full DID documents are now returned by the `bsky.social` login endpoint, which allows clients to connect directly to the back-end PDS if they want (for potentially better network latency)

## Will my stuff be impacted?

**Regular accounts using Bluesky App:** should be seamless. We expect that folks will not even need to re-log in unless something goes wrong.

**Client Applications:** same as the Bluesky App.

**Bots:** should be mostly unaffected; same as the Bluesky App.

**Feed Generators:** if verifying firehose repo signatures, will need to detect identity signing key changes (eg, via "cache busting"). Likewise, if verifying JWT auth token signatures, will need to detect identity changes. If subscribing to legacy PDS firehose, will need to switch to BGS firehose.

**Accounts (repos) with `did:web` DIDs:** we will probably need to coordinate with you to update DID documents, stay tuned. Updating to newer DID document formatting at the same time would be great: https://github.com/bluesky-social/atproto/discussions/1510

**Sandbox PDS operators:** we haven't pushed this round of code changes to the sandbox version of our PDS implementation. These changes will be disruptive when they do land, and will probably require a full wipe for most instances. The PDS reference implementation will only support sqlite (not postgresql), will use a distinct signing key for each account, etc.

**Independent PDS implementations:** none of these changes will be required by protocol. PDS operators will be free to keep the "account entryway" and "repo hosting" functionality combined in a single service, and can continue to use a single signing key for multiple accounts if they like. This is just a deployment pattern we have chosen to scale out our service for millions of accounts.

A follow-up comments to this post will tag some specific libraries and services which we have tested against.

## What is "identity cache busting"?

Federated services in atproto need to keep track of account identity metadata, including handle and DID document resolution. Clients can make requests to PDS instances for this information, but even they are likely to cache the results. Independent services will usually do direct resolution with persisted caching, similar to how DNS caching works.

Our current recommendation is that services cache identity metadata for between 8 and 24 hours. They may also simply rely on native DNS TTL and HTTP caching header information to determine an appropriate cache lifetime. Large services may want to "subscribe" to the DID PLC event log instead of re-fetching information for individual accounts.

In addition to this time-limited caching policy, we recommend automatically "busting" the cache when there is an indication that the identity may be stale. For example, if a repo signature or JWT signature fails to validate. A best practice is to have a separate shorter caching window for "bust" operations, so that failures do not result in a busy-loop of many rapid requests when an identity is in a broken state.

## Tell me more about this "virtual PDS" thing

More details coming soon, but this is mostly a mechanism to take advantage of the federated nature of the protocol to help us grow the Bluesky-hosted network beyond a single big database server. The "virtual PDS" handles some of the organization-wide "glue" tasks, like maintaining the `*.bsky.social` handle namespace, private account email addresses (which should be unique across all of our hosted accounts), and routing new account signups to a particular PDS instance.

We'll soon communicate more details of this architecture, and how clients can connect directly to backing PDS instances (for better network performance).

**Comment by bnewbold** [comment]
Pinging some specific folks about this change. As with previous pings, not trying to promote/exclude anybody; and sorry if this is spammy.

**Feed generators:** there are two parts of this: fetching feeds from a migrated account, and posts from a migrated account getting indexed. I have not seen any issues with *fetching* third-party feeds, but don't have enough data to verify that indexing is working for many (though no evidence that it is not). Tested "For You" (skygaze.io), "Blacksky" (@rudyfraser), "Catch Up" (skyfeed.xyz, @redsolver), "Home+", "Discover", "Likes" (@jcsalterego)

**Clients:** there is a lot of surface area to cover here, but tested briefly with GreySky (@mozzius) and seems functional.

**Libraries:** it looks like the great python library that @MarshalX maintains had an issue with additional fields included in JWTs. This is partially on us for making that particular change without thinking that it could impact downstream folks. Looks like there was a fast fix and upgrading fixes the problem.

**Comment by redsolver** [comment]
Quick question, when will the bsky.social PDS firehose stop including all events? One of my servers is still using that endpoint, so I would like to know if I need to do the switch today or tomorrow is enough, thanks!

**Reply by redsolver** [reply]
Just checked `@bnewbold.net`, new posts are not indexed, so I'll switch that indexer to `bsky.network` in about 10 hours and start from cursor `0`

**Reply by jazware** [reply]
`bsky.network` has a 3-day event retention so you'll get a lot of playback just FYI

**Reply by strideynet** [reply]
Yup - I suppose this is the kick I need to finally move FurryList over to BGS 😓

**Reply by tcurdt** [reply]
So to confirm: bsky.network will bring all traffic in one place?

Is a firehose even possible in a fully distributed future?

**Reply by snarfed** [reply]
It is! https://atproto.com/guides/overview#achieving-scale , https://blueskyweb.xyz/blog/5-5-2023-federation-architecture

**Comment by gildaswise** [comment]
[deck.blue](https://deck.blue)'s dev here

One thing that I need to test is the trick I'm using to show uploaded GIFs, we're just using the direct getBlob call and if that's broken between PDS instances, the trick won't work if any of the test accounts upload a .gif file

I'll keep an eye for any other issues and let you guys know! Also pinging @myConsciousness so he knows about this

**Reply by myConsciousness** [reply]
Perhaps related to @rudyfraser's thread

**Comment by rudyfraser** [comment]
Can you elaborate on the getBlob redirect? I use that for image classification right now. Do I need to do anything differently besides using the bsky.social endpoint?

**Reply by devinivy** [reply]
Yeah, check it out!  If you curl bsky.social for this blob, which has been migrated from bsky.social to morel.us-east.host.bsky.network, you'll get a 302 redirect and a location header indicating where the blob lives now:
```
$ curl 'https://bsky.social/xrpc/com.atproto.sync.getBlob?did=did:plc:ragtjsm2j2vknwkz3zp4oxrd&cid=bafkreic44ipbw6cbrp2skjcv2irfje4zhe3qnbnzspefogbadcn5qzsriy' -v
< HTTP/2 302 
< date: Tue, 07 Nov 2023 14:48:03 GMT
< content-type: application/json; charset=utf-8
< content-length: 68
< location: https://morel.us-east.host.bsky.network/xrpc/com.atproto.sync.getBlob?did=did:plc:ragtjsm2j2vknwkz3zp4oxrd&cid=bafkreic44ipbw6cbrp2skjcv2irfje4zhe3qnbnzspefogbadcn5qzsriy
...
```

If you follow the redirect to morel.us-east.host.bsky.network you'll get the blob itself:
```
curl 'https://morel.us-east.host.bsky.network/xrpc/com.atproto.sync.getBlob?did=did:plc:ragtjsm2j2vknwkz3zp4oxrd&cid=bafkreic44ipbw6cbrp2skjcv2irfje4zhe3qnbnzspefogbadcn5qzsriy'
```

So all you need to do is ensure your http client follows redirects.  If you happen to be looking-up the identity information for the owner of the blob in PLC, then you'll find that their PDS is morel.us-east.host.bsky.network, in which case you may be able to hit that PDS directly without any redirect.

**Reply by snarfed** [reply]
302 redirect response to an XRPC! Love it. Good sign that it'll hopefully be added to the officially supported XRPC response status codes: https://atproto.com/specs/xrpc#summary-of-http-status-codes , https://github.com/bluesky-social/atproto/discussions/1711#discussioncomment-7203160

**Reply by bnewbold** [reply]
To clarify one aspect, the redirects work if you are connecting to `bsky.social`. If you have a migrated account connected directly to a PDS (eg, morel), and you do a getBlob to that PDS (eg, morel) for another account, which hasn't migrated, I don't think that redirect will work.

In the short term you could stay connected to `bsky.social` for all accounts. With federation, you'll need to resolve repos to their home PDS (which could be anywhere), and do `getBlob` to that PDS, if you are fetching raw blobs.

This PDS resolution process is described a bit in a recent blog post, both for repo CAR exports and `getBlob` requests: https://atproto.com/blog/repo-export#on-another-instance

**Reply by imax9000** [reply]
Note that the `Authorization` header can get stripped during this redirect ([golang http client source](https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/net/http/client.go;l=987-1003)). In case of https://pkg.go.dev/net/http, it happens because it redirects to a different TLD (`bsky.social` -> `bsky.network`)

UPDATE: erm, my bad, the endpoint doesn't require authentication, but I'm using https://pkg.go.dev/golang.org/x/oauth2 which *does* add `Authentication` header to requests after the redirect, and PDS doesn't support `ES256K` signature.

**Reply by imax9000** [reply]
Out of curiosity, when federation happens, which auth token should the clients send to a potentially untrusted PDS?

**Reply by DavidBuchanan314** [reply]
JWTs have an "aud" (audience) field, that indicates what service the token is valid for. As long as you set the aud field appropriately, you can send the JWT to an untrusted PDS without concern, since the token would not be valid in any other context

**Reply by imax9000** [reply]
> JWTs have an "aud" (audience) field, that indicates what service the token is valid for. As long as you set the aud field appropriately, you can send the JWT to an untrusted PDS without concern, since the token would not be valid in any other context

Concern here is that a rogue PDS might use the token to hijack the account.

**Reply by DavidBuchanan314** [reply]
It can't use the token for anything else. If it tried to "replay" it to another service, the `aud` field would no longer match and it would be treated as invalid.

**Reply by imax9000** [reply]
It's the same token issued by the user's home server, right? The client treats it as opaque and doesn't modify in any way. So a rogue PDS can use it to impersonate the user to the PDS that hosts their account. Or am I missing something?

**Reply by DavidBuchanan314** [reply]
Ohhh I think I see what you mean. I don't think a trusted PDS should ever redirect to a non-trusted PDS. But, if there is a 302 redirect, I don't think you should ever send auth tokens to the redirect target in any case (the `net/http` behavior is what I'd intuitively expect). Those are both "I think" not "I know for sure", so yeah, I think it would be good to get some official guidance here

**Reply by jcsalterego** [reply]
It's always OK to send your auth tokens to Jerry's PDS Emporium

**Reply by imax9000** [reply]
Yup, I also left a comment to that end here https://github.com/golang/oauth2/issues/500#issuecomment-1805622318

For "normal" web services it's a bit less of an issue, since use cases for redirecting to a third-party host are carved out well enough. But in a federated environment it can bite your ass pretty hard.

I didn't find any specific requirements for removing headers on redirect in RFC7231, so client behaviour might vary :⁠-⁠\

**Reply by myConsciousness** [reply]
@redsolver @videah @gildaswise 

My understanding is that Dart's [http package](https://pub.dev/packages/http) handles redirects automatically. So since [my packages](https://github.com/myConsciousness/atproto.dart) also use the http package, we will not have to be aware of this redirect.

If you have any problems with my part of the package, please let me know.

**Comment by strideynet** [comment]
Hey - just wanted to clarify something as we've noticed an unexpected change in behaviour.

As part of our logic, when we onboard a new actor to the FurryList feeds, we fetch the latest commit cid for the user using `com.atproto.sync.getHead` and then we use `com.atproto.sync.getRecord` to fetch the `app.bsky.actor.profile` for that commit CID. This is because we use the commit cid as part of the primary key when storing profiles in our profile history table - we can't use the profile CID itself because if a user changes their profile, and then changes it back, this yields the same CID.

Since actors have started to be migrated to the new PDS, calling `com.atproto.sync.getHead` on bsky.social for those actors is giving:
```
{"error":"RepoNotFound","message":"Could not find root for DID: did:plc:2apayem2zzo4b72ppfjwz7bp"}
```

My understanding from reading this document was that we should expect most calls to bsky.social to continue working correctly. If this starting to fail is expected behaviour, should we instead be calling these RPCs on the BGS? The AppView? Or should we be determining the address of the PDS for that individual user and making the request directly there instead?

nb: I'm aware that we need to move off `getHead` and onto `getLatestCommit` but the behaviour seems the same for that as well.

Many thanks and apologies if I'm missing something here!

**Reply by jazware** [reply]
You should either be determining the PDS that the repo lives on before making a request to get a record from it (since that record won't be available on all PDSs) or make that request to the BGS since the BGS hosts a copy of all federated repos. I would recommend talking to the BGS out of simplicity for now in those cases. When fetching blobs right now you'll need to get the blobs from the PDS that hosts the repo since we don't have the blob mirror turned on on the BGS atm.

As for if this is expected behavior, I'm not sure, hopefully @devinivy can weigh in.

**Reply by strideynet** [reply]
Thanks for the advice - I think we'll switch this to the BGS so we support third party PDS down the line - but it is surprising to me that this is failing for the bsky.social "virtual pds" which I expected would forward this request to the right underlying PDS.

**Reply by jazware** [reply]
Are you making these requests with auth? I think the way the gateway works is by forwarding the request to _your_ PDS, not necessarily the PDS that holds the data you're looking for iirc. And for now by default it also responds to requests for data that lives on BSky.social for unauthenticated requests maybe.

**Reply by strideynet** [reply]
Unauthed - you can try it with curl:

```
noah@Noahs-MBP teleport % curl 'https://bsky.social/xrpc/com.atproto.sync.getHead?did=did:plc:2apayem2zzo4b72ppfjwz7bp'
{"error":"HeadNotFound","message":"Could not find root for DID: did:plc:2apayem2zzo4b72ppfjwz7bp"}%                                                                                                                                                                                                                  noah@Noahs-MBP teleport % 
```

That behaviour makes sense I suppose - we'll switch to the BGS regardless :) Thanks Jaz

**Reply by tolfdog** [reply]
It seems like the BGS and PDS have different behavior for `sync.getRecord` at a specific commit, namely the BGS can't get the record but the PDS can:

```
$ curl 'https://bsky.network/xrpc/com.atproto.sync.getRecord?did=did:plc:445avk3am7zpwlrj7aop746e&collection=app.bsky.actor.profile&rkey=self&commit=bafyreiewb3tgogeckkik62c2mxc3npxxuy4x4htzfilsdrcll3f2fulb6e'
{"error":"failed to get record from repo"}

$ curl 'https://bsky.social/xrpc/com.atproto.sync.getRecord?did=did:plc:445avk3am7zpwlrj7aop746e&collection=app.bsky.actor.profile&rkey=self&commit=bafyreiewb3tgogeckkik62c2mxc3npxxuy4x4htzfilsdrcll3f2fulb6e'
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.
```

**Reply by DavidBuchanan314** [reply]
This seems expected - you probably want to query the AppView (i.e. api.bsky.app) if you want a unified view of all repos. The BGS doesn't really store data beyond what it needs to do its job, iiuc, it just funnels stuff over to the AppView.

**Reply by tolfdog** [reply]
It looks like we don't have `sync.getRecord` on the AppView side:

```
$ curl 'https://api.bsky.app/xrpc/com.atproto.sync.getRecord?did=did:plc:445avk3am7zpwlrj7aop746e&collection=app.bsky.actor.profile&rkey=self&commit=bafyreidmnlnalgkbhyw2wupk4ixii5sx2gwzdclecrzaqt4t6y3m44gfwu'
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /xrpc/com.atproto.sync.getRecord</pre>
</body>
</html>
```

**Reply by jazware** [reply]
The BGS should be the canonical "network-level-view" of _repos_ (`com.atproto.sync`) while the AppView should be the same but for `app.bsky` namespaced stuff. That request _should_ work but I can't even seem to get it to download that record from `bsky.social`, I'm getting `Read error at byte 59` when trying to `wget` it or curl it, something wonky going on there.

**Reply by jazware** [reply]
Something seems to be up with that CID maybe? It looks like if you remove the `commit` from the request you get a record back both from the BGS and from the `bsky.social` though the one coming back from the PDS has a bunch of extra stuff in it.

**Reply by tolfdog** [reply]
here's a more self-contained way to do hit - with this it does give me the full record ok:

```
$ curl "https://bsky.social/xrpc/com.atproto.sync.getRecord?did=did:plc:445avk3am7zpwlrj7aop746e&collection=app.bsky.actor.profile&rkey=self&commit=$(curl 'https://bsky.social/xrpc/com.atproto.sync.getLatestCommit?did=did:plc:445avk3am7zpwlrj7aop746e' | jq -r '.cid')"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    91  100    91    0     0    355      0 --:--:-- --:--:-- --:--:--   365
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.

$ curl "https://bsky.network/xrpc/com.atproto.sync.getRecord?did=did:plc:445avk3am7zpwlrj7aop746e&collection=app.bsky.actor.profile&rkey=self&commit=$(curl 'https://bsky.network/xrpc/com.atproto.sync.getLatestCommit?did=did:plc:445avk3am7zpwlrj7aop746e' | jq -r '.cid')"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    91  100    91    0     0    349      0 --:--:-- --:--:-- --:--:--   358
{"error":"failed to get record from repo"}
```

**Reply by tolfdog** [reply]
hmm, trying this with your profile results in the same thing:

```
$ did=did:plc:q6gjnaw2blty4crticxkmujt
$ curl "https://morel.us-east.host.bsky.network/xrpc/com.atproto.sync.getRecord?did=$did&collection=app.bsky.actor.profile&rkey=self&commit=$(curl "https://morel.us-east.host.bsky.network/xrpc/com.atproto.sync.getLatestCommit?did=$did" | jq -r '.cid')"

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    91  100    91    0     0    297      0 --:--:-- --:--:-- --:--:--   304
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.

$ curl "https://bsky.network/xrpc/com.atproto.sync.getRecord?did=$did&collection=app.bsky.actor.profile&rkey=self&commit=$(curl "https://bsky.network/xrpc/com.atproto.sync.getLatestCommit?did=$did" | jq -r '.cid')"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    92  100    92    0     0    312      0 --:--:-- --:--:-- --:--:--   319
{"error":"failed to get record from repo"}
```

**Reply by dholms** [reply]
This seems to work from `bsky.network` if you don't include the commit ie
```
curl "https://bsky.network/xrpc/com.atproto.sync.getRecord?did=did:plc:445avk3am7zpwlrj7aop746e&collection=app.bsky.actor.profile&rkey=self"
```

The commit parameter is not needed and is a historical artifact of when repos maintained a commit history so it is safe to remove it from your request.

That being said we should:
- note that that parameter has been deprecated in the schema
- update BGS to just ignore the parameter instead of failing on the request

**Reply by jazware** [reply]
I think there may be a mixup of semantics between the BGS and PDS implementation then maybe? The `commit` being passed in that `getRecord` request to the BGS is being used to compare against the CID of the record you're looking up. The `getLatestCommit` is the most recent head of the repo though iirc.

**Reply by dholms** [reply]
As a rough rule of thumb:
- for `com.atproto.sync` routes for arbitrary repos, you should go to the BGS (at `bsky.network`)
- for `app.bsky` read routes, you should go to the appview (at `api.bsky.app`)
- for requests on behalf of a user, you should go to their pds or the vPDS (`bsky.social`/`morel.us-east1....`)
- To get canonical data pertaining to a repo, you should go to the users PDS (ie `morel.us-east1...`)

**Reply by jazware** [reply]
Yeah I think the BGS implementation of GetRecord has a bug with how it handles the provided commit at the moment. It's kinda weird though since there's no repo history any more, that `commit` param doesn't really make sense since we only have a copy of the most recent repo commit.

**Reply by tolfdog** [reply]
oh i see, we were originally using the getRecord endpoint to try get historical data for a user's profile - but if that's not possible then it sounds like we'll just have to avoid doing that on our side. thanks!

**Reply by strideynet** [reply]
> oh i see, we were originally using the getRecord endpoint to try get historical data for a user's profile

I'd say less trying to fetch historical profiles - but more just trying to establish a stable identifier for the current specific iteration of the profile (and correctly handle a profile flipping back and forth to some previous content). We don't really care about their profile history prior to us beginning to track them. Perhaps we can just use some arbitrary fake commit cid for the profile we fetch when we onboard them, and then use the real commit cids from over the firehose for any future updates.

**Reply by strideynet** [reply]
We could just call `getLatestCommit` and then call `getRecord` without the commit parameter, but I dislike the potential race there.

**Reply by jazware** [reply]
Could you use the CID of the record itself?

**Reply by strideynet** [reply]
That was my first guess, but unfortunately, if you change your profile back to something it was previously, the CID is then the same. It feels "incorrect" to me to not be able to distinguish between those two points in time. I may just have to bite the bullet on it and do that.

**Reply by jazware** [reply]
If you want a point-in-time thing you can use the `rev` of the repo as well but yeah the record itself doesn't have any kind of history and you'll end up with a race condition unless you get the whole repo.

**Reply by strideynet** [reply]
Hm - I think I actually figured out how to get the Commit CID from the results of `com.atproto.sync.getRecord`:

```go
package main

import (
	"bytes"
	"context"
	"fmt"
	"github.com/bluesky-social/indigo/api/atproto"
	"github.com/bluesky-social/indigo/repo"
	"github.com/bluesky-social/indigo/xrpc"
	"github.com/ipfs/go-datastore"
	blockstore "github.com/ipfs/go-ipfs-blockstore"
)

func main() {
	ctx := context.Background()
	x := &xrpc.Client{
		Host: "https://bsky.social",
	}
	myDID := "did:plc:dllwm3fafh66ktjofzxhylwk"

	latestCommitOut, err := atproto.SyncGetLatestCommit(ctx, x, myDID)
	if err != nil {
		panic(err)
	}

	buf := new(bytes.Buffer)
	params := map[string]interface{}{
		"collection": "app.bsky.actor.profile",
		"did":        myDID,
		"rkey":       "self",
	}
	if err := x.Do(ctx, xrpc.Query, "", "com.atproto.sync.getRecord", params, nil, buf); err != nil {
		panic(err)
	}
	data := buf.Bytes()

	bs := blockstore.NewBlockstore(datastore.NewMapDatastore())
	root, err := repo.IngestRepo(ctx, bs, bytes.NewReader(data))
	if err != nil {
		panic(err)
	}

	fmt.Println("Latest commit          : " + latestCommitOut.Cid)
	fmt.Println("Root from SyncGetRecord: " + root.String())
        // These two are the same
}
```

```
Latest commit          : bafyreifd3wxcj55mb4jt7mlgxcenrjeruksnegnvrvpryfo24w4jzdc5pe
Root from SyncGetRecord: bafyreifd3wxcj55mb4jt7mlgxcenrjeruksnegnvrvpryfo24w4jzdc5pe
```

This works fine until I point it at `bsky.network` instead - at which point I get an unexpected EOF when ingesting the Repo. So it's definitely feeling like `bsky.network` is returning something different to `bsky.social` for the same RPC

**Reply by strideynet** [reply]
Calling the `com.atproto.sync.getRecord` RPC against PDS and BGS gives two different amounts of data:
- PDS 4316 bytes
- BGS 502 bytes

The data returned by the BGS is not enough for `repo.IngestRepo` to properly build out the repo.

**Reply by strideynet** [reply]
Looking at the implementation of `(s *BGS) handleComAtprotoSyncGetRecord(` it feels like the BGS implementation of `com.atproto.sync.getRecord` is incorrect. It's returning just the record CBOR marshalled, rather than returning a `vnd.ipld.car` as the spec states.

**Comment by gtuckerkellogg** [comment]
We have an academic project that we think could be pretty exciting and I'l love to talk with someone at BlueSky about what federation would mean for it. The project would require running, and generating feeds for, thousands of accounts as a botnet. But it would be benevolent botnet. We were originally thinking of the bird app, but then Elon blew that up and took away "Twitter for Good", which would have been a natural partner. We've been considering our own mastadon server, which we know could work in principle, but Bluesky federation  might give us another option. Is there any way to set up a discussion on this? I don't want to sidetrack the discussion board for a niche conversation.

**Reply by bnewbold** [reply]
Sure, i'll send you an email.

**Reply by gtuckerkellogg** [reply]
Thanks. Just added my email address to my profile.
